NEW Use custom list for eagerloaded relations

GuySartorelli opened the pull request:
## Performance comparisons
The following comparison table was made using this code, where `Player` and `Team` are simple `DataObject` models, and `Team` has a `many_many` relation to `Player`.
The database has 29 `Team` records, each of which has 10 `Player` records.

```php
$eagerLoad = true;

$teams = Team::get();
if ($eagerLoad) {
    $teams = $teams->eagerLoad('Players');
}

memory_reset_peak_usage();

foreach ($teams as $team) {
    foreach ($team->Players() as $player) {
        echo ''; // effectively no-op
    }
}

var_dump(memory_get_peak_usage() / 1024 / 1024); // memory usage in MB
```
The times were recorded from the lekoala debugbar, which shows the total ms for db calls in the request. It necessarily includes some unrelated queries, but those queries are fixed so they don't affect the _difference_ between implementations.

To get the values, I ran the above code 3 times for each scenario, and the table reflects the averages of the values.

||total db query time|memory usage|
|---|---|---|
|no eager loading|16.21ms|0.96MB|
|eager loading before|7.95ms|1.44MB|
|eager loading after|7.69ms|1.23MB|


## Issue
- https://github.com/silverstripe/silverstripe-framework/issues/10864

## Out of scope
- https://github.com/silverstripe/silverstripe-framework/issues/5911
- https://github.com/silverstripe/silverstripe-framework/issues/10874

GuySartorelli committed:
NEW Use custom list for eagerloaded relations

GuySartorelli commented on src/ORM/ArrayLib.php:
```diff
@@ -274,4 +274,22 @@ public static function iterateVolatile(array &$list)
+     * Similar to shuffle, but retains the existing keys.
+     * Shuffles the array in place.
+     */
+    public static function shuffleAssociative(array &$array): void
```
This is needed because the keys of the new list's `$rows` array is associative, with IDs as keys. I could have just had it as a private method on the list itself but it seems sensible to put it here instead.

GuySartorelli commented on tests/php/ORM/EagerLoadedRelationListTest.php:
Most of these tests are pulled straight from `DataListTest`, and refactored to test the new class. A couple have been refactored to be easier to read e.g. using data providers, must most just been changed to test the new class.
Please don't ask me to change any of these to use data providers. I've already spent enough time making the tests nicer than they were.

Note that tests from the `DataListTest` class that test filtering/sorting by relations or using searchfilter syntax have been removed - see out of scope in PR description.

GuySartorelli commented on src/ORM/Connect/MySQLQuery.php:
```diff
@@ -64,11 +64,20 @@ public function getIterator(): Traversable
+    /**
+     * Rewind to the first row
+     */
+    public function rewind(): void
+    {
+        // Check for the method first since $this->handle is a mixed type
+        if (method_exists($this->handle, 'data_seek')) {
+            $this->handle->data_seek(0);
         }
     }
```
This (and the same change in `MySQLStatement`)  is required so we can break a loop in `DataList::addEagerLoadedDataToParent()` and then iterate over the result again later from the beginning

GuySartorelli commented on src/ORM/DataList.php:
```diff
@@ -930,50 +932,9 @@ private function setDataObjectEagerLoadedData(DataObject $item): void
-                }
+        foreach (array_keys($this->eagerLoadedData) as $relation) {
+            if (array_key_exists($itemID, $this->eagerLoadedData[$relation])) {
+                $item->setEagerLoadedData($relation, $this->eagerLoadedData[$relation][$itemID][$relation]);
```
I've restructured the `eagerLoadedData` array so that it only contains the _top_ level of relational data. Sub relations are now being added directly to the appropriate parent `EagerLoadedDataList` or `DataObject` record as part of the `fetchEagerLoadRelations()` implementation.

GuySartorelli commented on src/ORM/DataList.php:
```diff
@@ -1021,53 +982,52 @@ protected function getFinalisedQuery()
     }
 
-    private function getEagerLoadVariables(string $eagerLoadRelation): array
+    private function getEagerLoadVariables(string $relationChain, string $relationName, string $parentDataClass): array
```
This method now only fetches the variables for a given relation in the chain.

GuySartorelli commented on src/ORM/DataList.php:
```diff
@@ -1079,173 +1039,205 @@ private function executeQuery(): Query
 
     private function fetchEagerLoadRelations(Query $query): void
```
This method has been refactored so that:
1. Instead of having a full list of _partial_ relations in the chain (e.g. `relation1`, `relation1.relation2`, etc) it now only contains the _complete_ chains (e.g. `relation1.relation2`).
2. It loops over each relation in the chain, and _always_ passes the parent relation data into the various `fetchEagerLoad*()` methods
3. switch/case is used instead of chained if statements, as it is clearer to read

In turn, each of the `fetchEagerLoad*()` methods has been changed so that once they have fetched their data, they add it to the appropriate parent. For top level relations that's into the `eagerLoadedData` array. For subrelations that's the appropriate parent `EagerLoadedDataList` or `DataObject` record.

GuySartorelli commented on src/ORM/DataList.php:
```diff
@@ -1321,24 +1376,32 @@ private function fetchEagerLoadManyMany(
      */
     public function eagerLoad(...$relations): static
```
This method now stores a list of all partial and complete eagerload relation chains into `eagerLoadAllRelations` to avoid duplicates, and a list of all _complete_ eagerload relation chains into `eagerLoadRelationChains` which `fetchEagerLoadRelations()` iterates over.

GuySartorelli commented on src/ORM/DataList.php:
```diff
@@ -1310,35 +1376,43 @@ private function fetchEagerLoadManyMany(
+     * instead of a subclass of DataList such as HasManyList i.e. MyDataObject->MyHasManyRelation() returns an EagerLoadedDataList
      */
-    public function eagerLoad(...$relations): static
+    public function eagerLoad(...$relationChains): static
```
This method now stores a list of all partial and complete eagerload relation chains into `eagerLoadAllRelations` to avoid duplicates, and a list of all _complete_ eagerload relation chains into `eagerLoadRelationChains` which `fetchEagerLoadRelations()` iterates over.

GuySartorelli commented on src/ORM/EagerLoadedDataList.php:
```diff
@@ -0,0 +1,893 @@
+
+    /**
+     * ID-indexed array that holds the data from SQL queries for the list
+     * @var array<int,array>
+     */
+    private array $rows = [];
+
+    /**
+     * Nested eager-loaded data which applies to relations on records contained in this list
+     * @var array<int,self|DataObject>
```
Note that this `array<{keyType},{valueType}>` syntax is very useful for IDEs to declare both the type for the _key_ as well as the value. It isn't technically in [the PHPDoc spec](https://docs.phpdoc.org/guide/references/phpdoc/types.html#arrays) but seems to be widely supported and is even used by tools such as phpstan.
If there's hesitance to use this syntax I can change it to comply with the spec.

emteknetnz commented on src/ORM/EagerLoadedDataList.php:
```diff
@@ -0,0 +1,893 @@
+    public function getIDList(): array
+    {
+        $ids = $this->column('ID');
+        return array_combine($ids, $ids) ?: [];
```
```suggestion
        return array_combine($ids, $ids);
```
Is there any scenario where array_combine returns something falsey? https://www.php.net/manual/en/function.array-combine.php

GuySartorelli replied:
I feel like I did that based on something failing but I can't remember what, and I can't see how it _could_ return anything falsey (other than an empty array itself) so I'll remove that and if tests don't fail we'll call it good.

GuySartorelli replied:
Done

emteknetnz commented on src/ORM/EagerLoadedDataList.php:
```diff
@@ -0,0 +1,893 @@
+ * Note that when this list represents a relation, adding items to or removing items from this list will NOT
+ * affect the underlying relation in the database. This list is read-only.
+ */
+class EagerLoadedDataList extends ViewableData implements Relation, SS_List, Filterable, Sortable, Limitable
```
```suggestion
class EagerLoadedRelationList extends ViewableData implements Relation, SS_List, Filterable, Sortable, Limitable
```
Would this renaming the class to `EagerLoadedRelationList` make sense? That would do away with the need to make it mirror DataList.

Looking at this now it's confusing to me having the word 'DataList' in here since it's not a DataList, and infact it has a `private DataList $dataList` which its super weird - why would a DataList need a private DataList?

We could also take it further to just `EagerLoadedList` - this was briefly discussed early though rejected because it's less nice to say out loud :-)  However if we remove 'Relation' then we could probably just remove the ' implements Relation' bit and associated 3x methods? Perhaps that would be better?

GuySartorelli replied:
As discussed offline, `EagerLoadedList` is preferred, but we will retain `implements Relation` because 99% of the time that will be its use-case.

GuySartorelli replied:
Done

emteknetnz commented on src/ORM/EagerLoadedDataList.php:
```diff
@@ -0,0 +1,893 @@
+                $manyManyComponent['extraFields']
+            );
+        } else {
+            $this->dataList = $dataListClass::create($dataClass, '');
```
What's a scenario where a DataList that's not a RelationList is passed in for $dataListClass?

GuySartorelli replied:
It's used in testing, and could be used by dark wizards to similarly input data from who-knows-where to later hydrate some `DataObject` records.
The `$dataListClass` is used to hydrate records from data, so if the data passed in isn't for a relation, then the `$dataListClass` should just be a `DataList`.

emteknetnz commented on src/ORM/EagerLoadedDataList.php:
```diff
@@ -0,0 +1,893 @@
+            $this->extraFields = $manyManyComponent['extraFields'];
+        }
+    }
+
```
Missing a few methods from DataList that aren't here - though depending on whether we want to keep the word `DataList` in the class name will probably change what methods we want:

- addFilter
- addMany
- createSearchFilter
- newObject
- removeAll
- removeByFilter
- removeByID
- removeMany
- subtract (at very least want this given we have add(), sum() etc)

GuySartorelli replied:
IMO None of the following make sense in the context of eager-loaded data:
- removeAll
- removeByFilter
- removeByID
- removeMany
These are all above _removing_ data from the _database_ in the context of `DataList` and its subclasses.
In the case of `DataList`, all of these remove the actual records from the database.
In the case of the various relation lists, they remove the _relation_ between two records.
I think it is a much clearer experience for these methods to simply _not exist_ on this list, so calling them gives you an exception, and therefore you are aware that you were trying to do something this list doesn't support.

`addMany` doesn't make sense to implement here for a similar reason to the remove methods not being sensible - we don't add `DataObject` records, we add rows of data. So that method would already _not work the same_ as it does on `DataList`. I think it's better to just not implement it to make it clear that's not supported here.

Similarly, `newObject` doesn't make sense - the PHPDoc says "Return a new item to add to this DataList" but again we don't add `DataObject` records. It's a read-only list.

`createSearchFilter` is protected and therefore don't need to be implemented here

I will implement `addFilter` and `Subtract` for the sake of compatability.

GuySartorelli replied:
Done

emteknetnz commented on src/ORM/EagerLoadedDataList.php:
```diff
@@ -0,0 +1,893 @@
+
+        uasort($rows, function (array $row, array $other): int {
+            $compared = 0;
+            foreach ($this->sort as $column => $direction) {
```
should possibly use one of the natsort (natsort, natcasesort, strnatcmp, strnatcasecmp) methods here rather than straight <=> so that numbers get sorted better

https://www.php.net/manual/en/function.natsort.php

GuySartorelli replied:
What matters really is whether they're sorted the same way `DataList` sorts them. I've written a new test to verify this and you're right, there is a difference there between the way the db sorts these and the way this list currently does.
Will change the implementation appropriately so that the new test passes.

GuySartorelli replied:
It turns out that `<=>` already does natural sorting - but the database _doesn't_ do natural sorting on numeric strings. So I've had to treat numbers differently than numeric strings to match db behaviour.

emteknetnz commented on tests/php/ORM/ArrayLibTest.php:
```diff
@@ -335,4 +335,17 @@ public function testIterateVolatileModified()
+            $this->assertEquals($value, $copy[$key]);
+        }
+        // Check the order is different
+        $this->assertNotSame($list, $copy);
```
Is it possible that the shuffle randomly ends up with the same list as before? e.g shuffling an array [1,2] 50% of the time may end up back as [1,2] ?

GuySartorelli replied:
Possible but very unlikely. I don't think it's worth caring about but I can make it try a few times before failing if you want me to.

GuySartorelli replied:
I've actioned this, 'cause we do run CI a lot so it probably would have happened eventually. Just saves us a very little pain down the road.

emteknetnz commented on tests/php/ORM/DataListEagerLoadingTest.php:
```diff
@@ -696,7 +697,7 @@ private function iterateEagerLoadData(DataList $dataList, int $chunks = 0): arra
             }
             $s = ob_get_clean();
-            $s = preg_replace('/.*__START_ITERATE__/s', '', $s);
+            $s = preg_replace('/.*__START_ITERATE__/s', '', $s) ?? '';
```
```suggestion
            $s = preg_replace('/.*__START_ITERATE__/s', '', $s);
```
If an error occurred I'd prefer a hard error here than simply an empty string

GuySartorelli replied:
This was helpful in debugging - but we can always put this back if we need to debug stuff in the future so I'm fine with removing it.
Done.

emteknetnz commented on src/ORM/ArrayLib.php:
```diff
@@ -274,4 +274,21 @@ public static function iterateVolatile(array &$list)
     }
+
+    /**
+     * Similar to shuffle, but retains the existing keys.
```
`retains the existing keys.` is a bit confusing as the keys are what's being shuffled, not the values

GuySartorelli replied:
Would you prefer "retains the existing association between the keys and the values"? Or... something else that you suggest?

GuySartorelli replied:
Done

emteknetnz reviewed (changes requested):
There's a lot of changes here so I'll probably need to do 2 or 3 rounds of review

On extra thing to add / add a test for is that when attempting to filter by something with a dot e.g. `->filter(['MyRelation.Title' => 'lorem'])` - we should be throwing an exception so that the developer is aware this isn't possible

We currently have this in the corresponding docs pr
"
Note also that `EagerLoadedDataList` can't currently filter or sort by fields on relations using dot syntax (e.g. `sort('MySubRelation.Title')` won't work), nor filter using [`SearchFilter` syntax](/developer_guides/model/searchfilters/) (e.g. `filter('Title:StartsWith', 'Prefix')`).
"

GuySartorelli commented:
> On extra thing to add / add a test for is that when attempting to filter by something with a dot e.g. `->filter(['MyRelation.Title' => 'lorem'])` - we should be throwing an exception so that the developer is aware this isn't possible

Test added.

emteknetnz reviewed (changes requested):
Looks like there's some missing unit tests in EagerLoadedListTest.php

**EagerLoadedList public methods:**

### Has dedicated unit test
- [x] dataClass
- [x] getIterator
- [x] toNestedArray
- [x] map
- [x] each
- [x] max
- [x] min
- [x] avg
- [x] sum
- [x] canFilterBy
- [x] canSortBy
- [x] find
- [x] filter
- [x] filterAny
- [x] exclude
- [x] excludeAny
- [x] addFilter
- [x] subtract
- [x] filterByCallback
- [x] byID
- [x] byIDs
- [x] sort
- [x] shuffle
- [x] reverse
- [x] limit
- [x] has

### Tested in other tests
- [x] offsetGet
- [x] add
- [x] toArray
- [x] first
- [x] last
- [x] column
- [x] count

### Not tested
- [x] dbObject
- [x] getIDList
- [x] setByIDList
- [x] forForeignID
- [x] getRows
- [x] remove
- [x] columnUnique
- [x] debug
- [x] offsetExists
- [x] relation
- [x] createDataObject
- [x] getExtraData
- [x] getExtraFields
- [x] exists
- [x] offsetSet
- [x] offsetUnset

GuySartorelli commented:
I've now added tests for those.
For the ones that are implemented (i.e. don't just throw an exception) in the new list, I've added the test for both the new list _and_ for `DataList` (or the relevant subclass) which was also missing this coverage.

I've changed the implementation of `remove()` to more closely match `DataList`'s API, i.e. you pass in a `DataObject` record now just like you do with `DataList`.

I've changed the API for `add()` to match that of `DataList` - i.e. it expects a `DataObject` record - but it throws an exception, because there's no clean way to get _all_ of the data out into an array format to be stored correctly as 'rows' in this list type.
The old `add()` implementation has been moved to `addRow()`.

I've also improved the `relation()` implementation as I noticed while testing that there will be cases where we already have that data and therefore don't need to trigger a new db call.

Finally, I've swapped the order of arguments in the constructor - it is more intuitive to pass the dataclass first, I've found. This also means we can easily give a default for `$dataListClass` in the future if we want to.

emteknetnz commented on src/ORM/EagerLoadedList.php:
```diff
@@ -0,0 +1,1003 @@
+     * Removes a record from the list. Note that the record will not be removed from the
+     * database - this list is read-only.
+     */
+    public function remove($obj): static
```
```suggestion
    public function remove($item): static
```

Give it the same param name as DataList and add()

GuySartorelli replied:
Done

emteknetnz commented on src/ORM/EagerLoadedList.php:
```diff
@@ -0,0 +1,1003 @@
+        return $this->canFilterBy($fieldName);
+    }
+
+    public function find($key, $value): ?DataObject
```
```suggestion
    public function find(string $key, string $value): ?DataObject
```
We should strongly type all the things

While I can see why you didn't strongly type lots of args to match DataList, in this instance as well as many others you've added a strong return type which DataList doesn't have.  I think we should just strong type everything here as it's best practice.  Realistically getting the SS_List properly sorted so all the 3x list types match is a CMS 6 thing and that will involve adding strong typing to everything.

GuySartorelli replied:
> We should strongly type all the things

I disagree, and I _intentionally_ didn't add strong typing for parameters that _aren't_ strongly typed in `DataList`. If someone can pass a value into a method on `DataList`, they should be able to pass it in here as well. The specific example you've recommended changing here is a _great_ example - the `$value` might not be a string, it could be null or int or possibly other things, depending on what type of field we're looking at and how the database stores and returns that to us.

I don't think we should strongly type these parameters until we're able to commit to strongly typing them in `DataList` as well, so that the types will _always_ be valid or invalid across the board, and give the same errors/exceptions/experience for the developer. One of the main purposes of this list is to reduce the number of changes developers have to make to their existing codebase to take advantage of eagerloading.

Strong typing for return types just means that the _implementation_ of the method will return the same type, but doesn't affect what people _pass in_ which is the important part.

emteknetnz commented on src/ORM/EagerLoadedList.php:
```diff
@@ -0,0 +1,1003 @@
+    /**
+     * Check if this list has an item with the given ID
+     */
+    public function has(int $id): bool
```
This is a bit of a random method, I'd say probably just delete it since it's not on DataList.  You'd can just just use `byID()` and it for most normal use it'll take the same number of lines?

```php
if ($list->has(99)) {
    $obj = $list->byID(99);
    // do things
}

$obj = $list->byID(99);
if ($obj) {
  // do things
}
```

If you really want to keep it, `has()` isn't very obvious what it does, it should probably be `hasID()`

GuySartorelli replied:
`has()` was added in an earlier refactor of `DataList::fetchEagerLoadRelations()` where we needed to query the list to see if it had the parent record (when the data was still stored in a flat structure rather than a tree).

It looks like somewhere along the way I swapped it out for looping through `getRows()`... which was silly as it is less efficient _and_ required a `break 2` which is a little gross. I've now swapped that back to simply checking `has()` - and renamed that to `hasID()` per your request.

Using `byID()` is a lot less efficient, as it requires instantiating a dataobject, adding its nested eagerloading data, etc - when all we really need to do is check array_key_exists on the rows. Even if we change how we store the rows in the future, having `hasID()` loop over the rows and break when it finds the right one will _still_ be more efficient than instantiating the dataobject.

emteknetnz reviewed (changes requested):
Main thing is strongly type all the things

emteknetnz commented on src/ORM/EagerLoadedList.php:
```diff
@@ -0,0 +1,1003 @@
+    /**
+     * Return a new instance of the list with an added filter
+     */
+    public function addFilter(array $filterArray): static
```
```suggestion
    public function addFilter($filterArray): static
```
Match DataList

GuySartorelli replied:
In that case I need to not strongly type the `$filters` argument in `getMatches` - which, I _suppose_ someone might be passing in some other type of iterable here.
Done.

emteknetnz commented on src/ORM/EagerLoadedList.php:
```diff
@@ -0,0 +1,1003 @@
+        return $list;
+    }
+
+    private function standardiseColumn($column)
```
Strongly type

GuySartorelli replied:
Done

emteknetnz commented on src/ORM/EagerLoadedList.php:
```diff
@@ -0,0 +1,1003 @@
+        return $column;
+    }
+
+    private function validateSortColumn($column)
```
Strongly type

GuySartorelli replied:
Done

emteknetnz commented on src/ORM/EagerLoadedList.php:
```diff
@@ -0,0 +1,1003 @@
+     * Create a DataObject from the given SQL row.
+     * At a minimum, $row['ID'] must be set. Unsaved records cannot be eager loaded.
+     */
+    public function createDataObject(array $row): DataObject
```
```suggestion
    public function createDataObject($row): DataObject
```
Match DataList

GuySartorelli replied:
Done
